pkg/alertmanager/statefulset_test.go

```diff
@@ -918,81 +918,96 @@ func TestTerminationPolicy(t *testing.T) {
 	}
 }
 
-func TestClusterListenAddressForSingleReplica(t *testing.T) {
-	a := monitoringv1.Alertmanager{}
-	replicas := int32(1)
-	a.Spec.Version = operator.DefaultAlertmanagerVersion
-	a.Spec.Replicas = &replicas
-
-	statefulSet, err := makeStatefulSetSpec(&a, defaultTestConfig)
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	amArgs := statefulSet.Template.Spec.Containers[0].Args
-
-	containsEmptyClusterListenAddress := false
-
-	for _, arg := range amArgs {
-		if arg == "--cluster.listen-address=" {
-			containsEmptyClusterListenAddress = true
-		}
-	}
-
-	if !containsEmptyClusterListenAddress {
-		t.Fatal("expected stateful set to contain arg '--cluster.listen-address='")
-	}
-}
-
-func TestClusterListenAddressForSingleReplicaWithForceEnableClusterMode(t *testing.T) {
-	a := monitoringv1.Alertmanager{}
-	replicas := int32(1)
-	a.Spec.Version = operator.DefaultAlertmanagerVersion
-	a.Spec.Replicas = &replicas
-	a.Spec.ForceEnableClusterMode = true
-
-	statefulSet, err := makeStatefulSetSpec(&a, defaultTestConfig)
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	amArgs := statefulSet.Template.Spec.Containers[0].Args
-
-	containsEmptyClusterListenAddress := false
-
-	for _, arg := range amArgs {
-		if arg == "--cluster.listen-address=" {
-			containsEmptyClusterListenAddress = true
-		}
-	}
-
-	if containsEmptyClusterListenAddress {
-		t.Fatal("expected stateful set to not contain arg '--cluster.listen-address='")
-	}
-}
-
-func TestClusterListenAddressForMultiReplica(t *testing.T) {
-	a := monitoringv1.Alertmanager{}
-	replicas := int32(3)
-	a.Spec.Version = operator.DefaultAlertmanagerVersion
-	a.Spec.Replicas = &replicas
+func TestClusterListenAddress(t *testing.T) {
+
+	type testcase struct {
+		Name                   string
+		Replicas               int32
+		ClusterListenLocal     bool
+		ForceEnableClusterMode bool
+		ExpectedResult         string
+	}
+
+	testcases := []testcase{
+		{
+			Name:                   "ClusterListenAddressForSingleReplica",
+			Replicas:               1,
+			ForceEnableClusterMode: false,
+			ClusterListenLocal:     false,
+			ExpectedResult:         "--cluster.listen-address=",
+		},
+		{
+			Name:                   "ClusterListenAddressForSingleReplicaWithForceEnableClusterMode",
+			Replicas:               1,
+			ForceEnableClusterMode: true,
+			ClusterListenLocal:     false,
+			ExpectedResult:         "--cluster.listen-address=[$(POD_IP)]:9094",
+		},
+		{
+			Name:                   "ClusterListenAddressForSingleReplicaWithClusterListenLocal",
+			Replicas:               1,
+			ForceEnableClusterMode: false,
+			ClusterListenLocal:     true,
+			ExpectedResult:         "--cluster.listen-address=",
+		},
```

Comment:
For completeness, don't we miss an additional test case with replicas=1, ForceEnableClusterMode=true and ClusterListenLocal=true?

Reply:
Added testcase now in the latest commit.
https://github.com/rgynn/prometheus-operator/blob/2686a1cfc4c51b16db697ee25ab27afbf140b56e/pkg/alertmanager/statefulset_test.go#L973-L979